File: cp22/hpi-fun2/hexhex/executables/solution.py
import io
import os

primes = [
    2, 3, 5, 7,
    11, 13, 17, 19,
    23, 29, 31, 37,
    41, 43, 47, 53
]


left_bound = 0
right_bound = int(2e9)

max_depth = 0
w = 0


def dfs(x, d=0):
    if not (left_bound <= x <= right_bound):
        return False
    if x == w:
        return True
    if d >= max_depth:
        return False
    p = primes[d]
    return dfs(x + p, d + 1) or dfs(x - p, d + 1) or dfs(x * p, d + 1) or dfs(x // p, d + 1)


def solve():
    global max_depth, w
    h, w = input().split()
    h = int(h, 16)
    w = int(w, 16)

    for i in range(13):
        max_depth = i
        if dfs(h):
            print(i)
            return
    print(-1)


t = int(input())
for i in range(t):
    solve()

File: cp22/hpi-fun2/hexhex/executables/solution.cpp
// h.cpp -> h.exe

#include <bits/stdc++.h>
#define rep(i, n) for (int i = 0; i < (n); i++)
#define ll long long

using namespace std;

// clang-format off
int primes[] = {
    2, 3, 5, 7,
    11, 13, 17, 19,
    23, 29, 31, 37,
    41, 43, 47, 53,
};
// clang-format on

ll left_bound = -1e9;
ll right_bound = 1e9;

int w;
int max_depth;
bool _dfs(ll x, int d = 0) {
    if (left_bound > x || right_bound < x)
        return false;
    if (w == x)
        return true;
    if (d == max_depth)
        return false;
    ll p = primes[d];
    return _dfs(x + p, d + 1) || _dfs(x - p, d + 1) || _dfs(x * p, d + 1) ||
           _dfs(x / p, d + 1);
}

bool dfs(int x, int max_d) {
    max_depth = max_d;
    return _dfs(x);
}

void solve() {
    const int MAX_SOLUTION = 12;

    int h;
    cin >> hex >> h >> w;

    rep(i, MAX_SOLUTION + 1) {
        if (dfs(h, i)) {
            cout << i << "\n";
            return;
        }
    }
    cout << -1 << "\n";
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int t;
    cin >> t;

    rep(i, t) solve();

    return 0;
}